File: src/kernels/transport/measurement/simulate.zig
//! Purpose:
//!   Materialize measurement-space radiance, irradiance, reflectance, and
//!   optional Jacobian products from prepared transport input.
//!
//! Physics:
//!   Runs the transport executor across the resolved spectral grid, applies
//!   instrument integration and slit convolution, and reduces the results into
//!   summary and product-level outputs.
//!
//! Vendor:
//!   `measurement simulation` stage
//!
//! Design:
//!   The transport executor is kept separate from the measurement reduction so
//!   the same prepared state can feed summary and full-product materialization.
//!
//! Invariants:
//!   The transport workspace must be shape-compatible with the scene's sample
//!   count and the resolved transport layer count.
//!
//! Validation:
//!   Measurement-space summary and product tests.

const std = @import("std");
const SpectralChannel = @import("../../../model/Instrument.zig").SpectralChannel;
const Scene = @import("../../../model/Scene.zig").Scene;
const OpticsPreparation = @import("../../optics/preparation.zig");
const calibration = @import("../../spectra/calibration.zig");
const convolution = @import("../../spectra/convolution.zig");
const grid = @import("../../spectra/grid.zig");
const common = @import("../common.zig");
const Postprocess = @import("postprocess.zig");
const SamplePlan = @import("sample_plan.zig");
const SpectralEval = @import("spectral_eval.zig");
const Types = @import("types.zig");
const Workspace = @import("workspace.zig");

const Allocator = std.mem.Allocator;
const max_summary_samples: u32 = 128;

/// Purpose:
///   Execute the transport solver across the scene spectral grid.
///
/// Physics:
///   Materializes radiance, irradiance, reflectance, and optional Jacobian
///   arrays after route-specific integration and calibration.
///
/// Inputs:
///   `scene` defines the resolved spectral grid, `route` selects the transport
///   family, `prepared` carries the optical state, and `buffers` provides the
///   reusable scratch slices.
///
/// Outputs:
///   Returns measurement-space summary statistics and fills the product
///   buffers in place.
///
/// Validation:
///   Measurement-space summary and product tests.
pub fn simulateInternal(
    allocator: Allocator,
    scene: *const Scene,
    route: common.Route,
    prepared: *const OpticsPreparation.PreparedOpticalState,
    providers: Types.ProviderBindings,
    buffers: Workspace.Buffers,
    evaluation_cache: *SpectralEval.SpectralEvaluationCache,
) Workspace.Error!Types.MeasurementSpaceSummary {
    try scene.validate();
    const sample_count: usize = @intCast(scene.spectral_grid.sample_count);
    try Workspace.validateBuffers(sample_count, buffers);

    const spectral_grid: grid.SpectralGrid = .{
        .start_nm = scene.spectral_grid.start_nm,
        .end_nm = scene.spectral_grid.end_nm,
        .sample_count = scene.spectral_grid.sample_count,
    };
    const resolved_axis: grid.ResolvedAxis = .{
        .base = spectral_grid,
        .explicit_wavelengths_nm = scene.observation_model.measured_wavelengths_nm,
    };
    try resolved_axis.validate();

    const radiance_calibration = providers.instrument.calibrationForScene(scene, .radiance);
    const irradiance_calibration = providers.instrument.calibrationForScene(scene, .irradiance);
    const radiance_slit_kernel = providers.instrument.slitKernelForScene(scene, .radiance);
    const irradiance_slit_kernel = providers.instrument.slitKernelForScene(scene, .irradiance);
    const uses_integrated_radiance_sampling = providers.instrument.usesIntegratedSampling(scene, .radiance);
    const uses_integrated_irradiance_sampling = providers.instrument.usesIntegratedSampling(scene, .irradiance);
    const span_nm = scene.spectral_grid.end_nm - scene.spectral_grid.start_nm;
    const safe_span = if (span_nm <= 0.0) 1.0 else span_nm;
    const sample_plans = try SamplePlan.buildSamplePlans(
        allocator,
        scene,
        prepared,
        &resolved_axis,
        radiance_calibration,
        irradiance_calibration,
        providers,
    );
    defer allocator.free(sample_plans);
    const forward_misses = try SamplePlan.collectUniqueForwardMisses(
        allocator,
        sample_plans,
    );
    defer allocator.free(forward_misses);
    try SpectralEval.prefetchForwardSamples(
        allocator,
        scene,
        route,
        prepared,
        providers,
        safe_span,
        forward_misses,
        evaluation_cache,
    );

    var radiance_sum: f64 = 0.0;
    var irradiance_sum: f64 = 0.0;
    var reflectance_sum: f64 = 0.0;
    var noise_sum: f64 = 0.0;
    var jacobian_sum: f64 = 0.0;
    const transport_layer_count = Workspace.resolvedTransportLayerCount(route, prepared);
    if (buffers.layer_inputs.len < transport_layer_count or
        buffers.source_interfaces.len < transport_layer_count + 1 or
        buffers.rtm_quadrature_levels.len < transport_layer_count + 1 or
        buffers.pseudo_spherical_level_starts.len < transport_layer_count + 1)
    {
        return error.ShapeMismatch;
    }

    for (sample_plans, 0..) |plan, index| {
        const nominal_wavelength_nm = plan.nominal_wavelength_nm;
        buffers.wavelengths[index] = nominal_wavelength_nm;

        const integrated = try SpectralEval.integrateForwardAtNominal(
            allocator,
            scene,
            route,
            prepared,
            plan.radiance_wavelength_nm,
            safe_span,
            providers,
            buffers.layer_inputs[0..transport_layer_count],
            buffers.pseudo_spherical_layers,
            buffers.source_interfaces[0 .. transport_layer_count + 1],
            buffers.rtm_quadrature_levels[0 .. transport_layer_count + 1],
            buffers.pseudo_spherical_samples,
            buffers.pseudo_spherical_level_starts[0 .. transport_layer_count + 1],
            buffers.pseudo_spherical_level_altitudes[0 .. transport_layer_count + 1],
            evaluation_cache,
            &plan.radiance_integration,
        );
        buffers.scratch[index] = integrated.radiance;
        if (buffers.jacobian) |jacobian| jacobian[index] = integrated.jacobian;
    }
    if (uses_integrated_radiance_sampling) {
        // DECISION:
        //   Integrated sampling bypasses slit convolution because the
        //   instrument already performed the spectral integration.
        @memcpy(buffers.radiance, buffers.scratch);
    } else {
        try convolution.apply(buffers.scratch, radiance_slit_kernel[0..], buffers.radiance);
    }
    try Postprocess.applyChannelCorrections(
        scene,
        .radiance,
        radiance_calibration,
        prepared.depolarization_factor,
        buffers.wavelengths,
        buffers.radiance,
        buffers.scratch_aux,
    );
    for (sample_plans, 0..) |plan, index| {
        buffers.scratch[index] = try SpectralEval.integrateIrradianceAtNominal(
            scene,
            prepared,
            plan.irradiance_wavelength_nm,
            safe_span,
            evaluation_cache,
            &plan.irradiance_integration,
        );
    }
    if (uses_integrated_irradiance_sampling) {
        @memcpy(buffers.irradiance, buffers.scratch);
    } else {
        try convolution.apply(buffers.scratch, irradiance_slit_kernel[0..], buffers.irradiance);
    }
    try Postprocess.applyChannelCorrections(
        scene,
        .irradiance,
        irradiance_calibration,
        prepared.depolarization_factor,
        buffers.wavelengths,
        buffers.irradiance,
        buffers.scratch_aux,
    );
    try calibration.applyRingSpectrum(
        scene.observation_model.resolvedRingControls(),
        buffers.wavelengths,
        buffers.irradiance,
        buffers.radiance,
        buffers.scratch_aux,
    );
    const solar_cosine = scene.geometry.solarCosineAtAltitude(0.0);
    for (0..sample_count) |index| {
        buffers.reflectance[index] = (buffers.radiance[index] * std.math.pi) /
            @max(buffers.irradiance[index] * solar_cosine, 1e-9);
        radiance_sum += buffers.radiance[index];
        irradiance_sum += buffers.irradiance[index];
        reflectance_sum += buffers.reflectance[index];
    }

    const radiance_noise_sigma = if (buffers.radiance_noise_sigma) |sigma|
        sigma
    else if (buffers.noise_sigma) |sigma|
        sigma
    else if (buffers.reflectance_noise_sigma != null)
        buffers.scratch
    else
        null;
    if (radiance_noise_sigma) |sigma| {
        try Postprocess.materializeChannelSigma(providers, scene, .radiance, buffers.wavelengths, buffers.radiance, sigma);
    }
    if (buffers.noise_sigma) |noise_sigma| {
        const sigma = radiance_noise_sigma orelse return error.ShapeMismatch;
        if (noise_sigma.ptr != sigma.ptr) {
            @memcpy(noise_sigma, sigma);
        }
    }

    const irradiance_noise_sigma = if (buffers.irradiance_noise_sigma) |sigma|
        sigma
    else if (buffers.reflectance_noise_sigma != null)
        buffers.scratch_aux
    else
        null;
    if (irradiance_noise_sigma) |sigma| {
        try Postprocess.materializeChannelSigma(providers, scene, .irradiance, buffers.wavelengths, buffers.irradiance, sigma);
    }

    if (buffers.reflectance_noise_sigma) |reflectance_noise_sigma| {
        const radiance_sigma = radiance_noise_sigma orelse return error.ShapeMismatch;
        const irradiance_sigma = irradiance_noise_sigma orelse return error.ShapeMismatch;
        for (0..sample_count) |index| {
            const radiance_term = if (radiance_sigma.len == sample_count and buffers.radiance[index] > 0.0)
                buffers.reflectance[index] * (radiance_sigma[index] / @max(buffers.radiance[index], 1.0e-12))
            else
                0.0;
            const irradiance_term = if (irradiance_sigma.len == sample_count and buffers.irradiance[index] > 0.0)
                buffers.reflectance[index] * (irradiance_sigma[index] / @max(buffers.irradiance[index], 1.0e-12))
            else
                0.0;
            reflectance_noise_sigma[index] = std.math.sqrt(radiance_term * radiance_term + irradiance_term * irradiance_term);
        }
        try calibration.applyReflectanceCalibrationErrorSigma(
            scene.observation_model.resolvedReflectanceCalibration(),
            buffers.wavelengths,
            buffers.reflectance,
            reflectance_noise_sigma,
            buffers.scratch_aux,
        );
    }

    if (radiance_noise_sigma) |sigma| {
        for (sigma) |value| noise_sum += value;
    }

    var mean_jacobian: ?f64 = null;
    if (buffers.jacobian) |jacobian| {
        if (!uses_integrated_radiance_sampling) {
            try convolution.apply(jacobian, radiance_slit_kernel[0..], buffers.scratch);
            @memcpy(jacobian, buffers.scratch);
        }
        try Postprocess.applyChannelJacobianCorrections(
            scene,
            .radiance,
            radiance_calibration,
            prepared.depolarization_factor,
            buffers.wavelengths,
            jacobian,
            buffers.scratch_aux,
        );
        // DECISION:
        //   Ring synthesis uses the irradiance-only basis from the current
        //   forward model, so it does not change the routed radiance Jacobian.
        for (jacobian) |value| jacobian_sum += value;
        mean_jacobian = jacobian_sum / @as(f64, @floatFromInt(sample_count));
    }
    return .{
        .sample_count = @intCast(sample_count),
        .wavelength_start_nm = buffers.wavelengths[0],
        .wavelength_end_nm = buffers.wavelengths[sample_count - 1],
        .mean_radiance = radiance_sum / @as(f64, @floatFromInt(sample_count)),
        .mean_irradiance = irradiance_sum / @as(f64, @floatFromInt(sample_count)),
        .mean_reflectance = reflectance_sum / @as(f64, @floatFromInt(sample_count)),
        .mean_noise_sigma = if (radiance_noise_sigma != null)
            noise_sum / @as(f64, @floatFromInt(sample_count))
        else
            0.0,
        .mean_jacobian = mean_jacobian,
    };
}

pub fn simulate(
    allocator: Allocator,
    scene: *const Scene,
    route: common.Route,
    prepared: *const OpticsPreparation.PreparedOpticalState,
    providers: Types.ProviderBindings,
    buffers: Workspace.Buffers,
) Workspace.Error!Types.MeasurementSpaceSummary {
    var evaluation_cache = SpectralEval.SpectralEvaluationCache.init(allocator);
    defer evaluation_cache.deinit();
    evaluation_cache.reset();
    return simulateInternal(allocator, scene, route, prepared, providers, buffers, &evaluation_cache);
}

/// Purpose:
///   Materialize a summary-only measurement-space product.
pub fn simulateSummary(
    allocator: Allocator,
    scene: *const Scene,
    route: common.Route,
    prepared: *const OpticsPreparation.PreparedOpticalState,
    providers: Types.ProviderBindings,
) Workspace.Error!Types.MeasurementSpaceSummary {
    var workspace: Workspace.SummaryWorkspace = .{};
    defer workspace.deinit(allocator);
    return simulateSummaryWithWorkspace(allocator, &workspace, scene, route, prepared, providers);
}

/// Purpose:
///   Materialize a summary-only measurement-space product with reusable buffers.
pub fn simulateSummaryWithWorkspace(
    allocator: Allocator,
    workspace: *Workspace.SummaryWorkspace,
    scene: *const Scene,
    route: common.Route,
    prepared: *const OpticsPreparation.PreparedOpticalState,
    providers: Types.ProviderBindings,
) Workspace.Error!Types.MeasurementSpaceSummary {
    var summary_scene = scene.*;
    // GOTCHA:
    //   Summary mode truncates very long spectral grids so it can stay
    //   lightweight while preserving the full-product path for complete runs.
    if (summary_scene.spectral_grid.sample_count > max_summary_samples) {
        summary_scene.spectral_grid.sample_count = max_summary_samples;
    }
    return simulateInternal(
        allocator,
        &summary_scene,
        route,
        prepared,
        providers,
        try workspace.buffers(allocator, &summary_scene, route, providers),
        try workspace.spectralCache(allocator),
    );
}

File: rust_src/forward_model/instrument_grid/grid_calculation/simulate.rs
use crate::{
    common::errors,
    forward_model::{
        implementations::root as implementation_root,
        instrument_grid::{
            grid_calculation::{
                cache::SpectralEvaluationCache,
                forward_input::ForwardInputBuffers,
                postprocess,
                spectral_eval::{
                    self, integrate_forward_at_nominal, integrate_irradiance_at_nominal,
                },
                spectral_forward, storage,
                types::{Implementations, InstrumentGridProduct, InstrumentGridSummary},
                wavelength_plan::WavelengthSampling,
                wavelength_sampling,
            },
            spectral_math::{
                calibration,
                grid::{self, ResolvedAxis, SpectralGrid},
            },
        },
        jacobian,
        optical_properties::state_build::PreparedOpticalState,
        radiative_transfer::common_types::{
            LayerInput, PseudoSphericalSample, Route, RtmQuadratureLevel, SourceInterfaceInput,
        },
    },
    input::{instrument::SpectralChannel, scene::Scene},
};
use rayon::prelude::*;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Error {
    Scene(errors::Error),
    Grid(grid::Error),
    SpectralEval(spectral_eval::Error),
    InstrumentIntegration(crate::forward_model::implementations::instrument::Error),
    Postprocess(postprocess::Error),
    WavelengthSampling(wavelength_sampling::Error),
}

impl From<errors::Error> for Error {
    fn from(value: errors::Error) -> Self {
        Self::Scene(value)
    }
}

impl From<grid::Error> for Error {
    fn from(value: grid::Error) -> Self {
        Self::Grid(value)
    }
}

impl From<spectral_eval::Error> for Error {
    fn from(value: spectral_eval::Error) -> Self {
        Self::SpectralEval(value)
    }
}

impl From<crate::forward_model::implementations::instrument::Error> for Error {
    fn from(value: crate::forward_model::implementations::instrument::Error) -> Self {
        Self::InstrumentIntegration(value)
    }
}

impl From<wavelength_sampling::Error> for Error {
    fn from(value: wavelength_sampling::Error) -> Self {
        Self::WavelengthSampling(value)
    }
}

impl From<postprocess::Error> for Error {
    fn from(value: postprocess::Error) -> Self {
        Self::Postprocess(value)
    }
}

struct RunningSummary {
    radiance_sum: f64,
    irradiance_sum: f64,
    reflectance_sum: f64,
    noise_sum: f64,
    jacobian_sum: jacobian::Vector,
}

struct ForwardScratch {
    layer_inputs: Vec<LayerInput>,
    pseudo_spherical_layers: Vec<LayerInput>,
    source_interfaces: Vec<SourceInterfaceInput>,
    rtm_quadrature_levels: Vec<RtmQuadratureLevel>,
    pseudo_spherical_samples: Vec<PseudoSphericalSample>,
    pseudo_spherical_level_starts: Vec<usize>,
    pseudo_spherical_level_altitudes: Vec<f64>,
}

impl ForwardScratch {
    fn new(transport_layer_count: usize, pseudo_spherical_sample_count: usize) -> Self {
        Self {
            layer_inputs: vec![LayerInput::default(); transport_layer_count],
            pseudo_spherical_layers: vec![LayerInput::default(); pseudo_spherical_sample_count],
            source_interfaces: vec![SourceInterfaceInput::default(); transport_layer_count + 1],
            rtm_quadrature_levels: vec![RtmQuadratureLevel::default(); transport_layer_count + 1],
            pseudo_spherical_samples: vec![
                PseudoSphericalSample::default();
                pseudo_spherical_sample_count
            ],
            pseudo_spherical_level_starts: vec![0; transport_layer_count + 1],
            pseudo_spherical_level_altitudes: vec![0.0; transport_layer_count + 1],
        }
    }

    fn buffers(&mut self) -> ForwardInputBuffers<'_> {
        ForwardInputBuffers {
            layer_inputs: &mut self.layer_inputs,
            pseudo_spherical_layers: &mut self.pseudo_spherical_layers,
            source_interfaces: &mut self.source_interfaces,
            rtm_quadrature_levels: &mut self.rtm_quadrature_levels,
            pseudo_spherical_samples: &mut self.pseudo_spherical_samples,
            pseudo_spherical_level_starts: &mut self.pseudo_spherical_level_starts,
            pseudo_spherical_level_altitudes: &mut self.pseudo_spherical_level_altitudes,
        }
    }
}

impl RunningSummary {
    fn new() -> Self {
        Self {
            radiance_sum: 0.0,
            irradiance_sum: 0.0,
            reflectance_sum: 0.0,
            noise_sum: 0.0,
            jacobian_sum: jacobian::zero(),
        }
    }

    fn to_summary(
        &self,
        wavelengths: &[f64],
        has_noise_sigma: bool,
        mean_jacobian: Option<jacobian::Vector>,
    ) -> InstrumentGridSummary {
        let denominator = wavelengths.len() as f64;
        InstrumentGridSummary {
            sample_count: wavelengths.len() as u32,
            wavelength_start_nm: wavelengths[0],
            wavelength_end_nm: wavelengths[wavelengths.len() - 1],
            mean_radiance: self.radiance_sum / denominator,
            mean_irradiance: self.irradiance_sum / denominator,
            mean_reflectance: self.reflectance_sum / denominator,
            mean_noise_sigma: if has_noise_sigma {
                self.noise_sum / denominator
            } else {
                0.0
            },
            mean_jacobian,
        }
    }
}

fn precompute_forward_cache(
    scene: &Scene,
    route: Route,
    prepared: &PreparedOpticalState,
    wavelength_sampling: &[WavelengthSampling],
    cache: &mut SpectralEvaluationCache,
) -> Result<(), spectral_eval::Error> {
    let forward_misses = wavelength_sampling::collect_unique_forward_misses(wavelength_sampling);
    let transport_layer_count = storage::resolved_transport_layer_count(route, prepared);
    let pseudo_spherical_sample_count =
        storage::resolved_pseudo_spherical_sample_count(scene, route, prepared);

    // Most O2 A instrument samples overlap heavily across nominal wavelengths.
    // Computing each unique forward wavelength once avoids duplicate RTM work,
    // and thread-local buffers keep the parallel path free of shared mutation.
    let samples: Result<Vec<_>, spectral_eval::Error> = forward_misses
        .par_iter()
        .map_init(
            || ForwardScratch::new(transport_layer_count, pseudo_spherical_sample_count),
            |scratch, miss| {
                spectral_forward::compute_forward_sample_at_wavelength(
                    scene,
                    route,
                    prepared,
                    miss.wavelength_nm,
                    scratch.buffers(),
                )
                .map(|sample| (miss.wavelength_nm, sample))
                .map_err(spectral_eval::Error::from)
            },
        )
        .collect();

    for (wavelength_nm, sample) in samples? {
        cache.put_forward(wavelength_nm, sample);
    }

    Ok(())
}

pub fn simulate_product(
    scene: &Scene,
    route: Route,
    prepared: &PreparedOpticalState,
) -> Result<InstrumentGridProduct, Error> {
    simulate_product_with_implementations(scene, route, prepared, implementation_root::exact())
}

pub fn simulate_product_with_implementations(
    scene: &Scene,
    route: Route,
    prepared: &PreparedOpticalState,
    implementations: Implementations,
) -> Result<InstrumentGridProduct, Error> {
    scene.validate()?;
    let axis = ResolvedAxis {
        base: SpectralGrid {
            start_nm: scene.spectral_grid.start_nm,
            end_nm: scene.spectral_grid.end_nm,
            sample_count: scene.spectral_grid.sample_count,
        },
        explicit_wavelengths_nm: scene.observation_model.measured_wavelengths_nm.clone(),
    };
    axis.validate()?;
    let instrument_provider = implementations.instrument;
    let noise_provider = implementations.noise;
    let wavelength_sampling = wavelength_sampling::build_wavelength_sampling(
        scene,
        &axis,
        prepared,
        instrument_provider,
    )?;
    let mut cache = SpectralEvaluationCache::default();
    precompute_forward_cache(scene, route, prepared, &wavelength_sampling, &mut cache)?;

    let sample_count = scene.spectral_grid.sample_count as usize;
    let transport_layer_count = storage::resolved_transport_layer_count(route, prepared);
    let pseudo_spherical_sample_count =
        storage::resolved_pseudo_spherical_sample_count(scene, route, prepared);
    let mut layer_inputs = vec![LayerInput::default(); transport_layer_count];
    let mut pseudo_spherical_layers = vec![LayerInput::default(); pseudo_spherical_sample_count];
    let mut source_interfaces = vec![SourceInterfaceInput::default(); transport_layer_count + 1];
    let mut rtm_quadrature_levels = vec![RtmQuadratureLevel::default(); transport_layer_count + 1];
    let mut pseudo_spherical_samples =
        vec![PseudoSphericalSample::default(); pseudo_spherical_sample_count];
    let mut pseudo_spherical_level_starts = vec![0; transport_layer_count + 1];
    let mut pseudo_spherical_level_altitudes = vec![0.0; transport_layer_count + 1];

    let mut wavelengths = vec![0.0; sample_count];
    let mut radiance = vec![0.0; sample_count];
    let mut irradiance = vec![0.0; sample_count];
    let mut reflectance = vec![0.0; sample_count];
    let mut noise_sigma = vec![0.0; sample_count];
    let mut scratch = vec![0.0; sample_count];
    let mut scratch_aux = vec![0.0; sample_count];
    let wants_jacobian = route.derivative_mode != crate::input::scene::DerivativeMode::None;
    let mut jacobian_values =
        wants_jacobian.then(|| vec![0.0; sample_count * jacobian::STATE_COUNT]);
    let mut summary = RunningSummary::new();
    let solar_cosine = scene.geometry.solar_cosine_at_altitude(0.0);
    let mut forward_buffers = ForwardInputBuffers {
        layer_inputs: &mut layer_inputs,
        pseudo_spherical_layers: &mut pseudo_spherical_layers,
        source_interfaces: &mut source_interfaces,
        rtm_quadrature_levels: &mut rtm_quadrature_levels,
        pseudo_spherical_samples: &mut pseudo_spherical_samples,
        pseudo_spherical_level_starts: &mut pseudo_spherical_level_starts,
        pseudo_spherical_level_altitudes: &mut pseudo_spherical_level_altitudes,
    };

    for (index, plan) in wavelength_sampling.iter().enumerate() {
        let sample = integrate_forward_at_nominal(
            scene,
            route,
            prepared,
            plan.radiance_wavelength_nm,
            &mut forward_buffers,
            &mut cache,
            &plan.radiance_integration,
        )?;
        let sample_irradiance = integrate_irradiance_at_nominal(
            scene,
            plan.irradiance_wavelength_nm,
            &mut cache,
            &plan.irradiance_integration,
        )?;
        wavelengths[index] = plan.nominal_wavelength_nm;
        radiance[index] = sample.radiance;
        irradiance[index] = sample_irradiance;
        if let Some(values) = &mut jacobian_values {
            let start = index * jacobian::STATE_COUNT;
            values[start..start + jacobian::STATE_COUNT].copy_from_slice(&sample.jacobian);
        }
    }

    postprocess::apply_channel_corrections(
        scene,
        SpectralChannel::Radiance,
        (instrument_provider.calibration_for_scene)(scene, SpectralChannel::Radiance),
        prepared.depolarization_factor,
        &wavelengths,
        &mut radiance,
        &mut scratch,
    )?;
    postprocess::apply_channel_corrections(
        scene,
        SpectralChannel::Irradiance,
        (instrument_provider.calibration_for_scene)(scene, SpectralChannel::Irradiance),
        prepared.depolarization_factor,
        &wavelengths,
        &mut irradiance,
        &mut scratch,
    )?;
    calibration::apply_ring_spectrum(
        &scene.observation_model.resolved_ring_controls(),
        &wavelengths,
        &irradiance,
        &mut radiance,
        &mut scratch,
    )
    .map_err(postprocess::Error::from)?;

    if let Some(values) = &mut jacobian_values {
        for state_index in 0..jacobian::STATE_COUNT {
            for sample_index in 0..sample_count {
                scratch[sample_index] = values[sample_index * jacobian::STATE_COUNT + state_index];
            }
            postprocess::apply_channel_jacobian_corrections(
                scene,
                SpectralChannel::Radiance,
                (instrument_provider.calibration_for_scene)(scene, SpectralChannel::Radiance),
                prepared.depolarization_factor,
                &wavelengths,
                &mut scratch,
                &mut scratch_aux,
            )?;
            for sample_index in 0..sample_count {
                values[sample_index * jacobian::STATE_COUNT + state_index] = scratch[sample_index];
            }
        }
    }

    let has_noise_sigma = (noise_provider.materializes_sigma)(scene, SpectralChannel::Radiance);
    postprocess::materialize_channel_sigma(
        noise_provider,
        scene,
        SpectralChannel::Radiance,
        &wavelengths,
        &radiance,
        &mut noise_sigma,
    )?;
    summary.noise_sum = noise_sigma.iter().sum();

    for index in 0..sample_count {
        let sample_reflectance = if irradiance[index] > 0.0 && solar_cosine > 0.0 {
            radiance[index] * std::f64::consts::PI / (irradiance[index] * solar_cosine)
        } else {
            0.0
        };
        reflectance[index] = sample_reflectance;
        summary.radiance_sum += radiance[index];
        summary.irradiance_sum += irradiance[index];
        summary.reflectance_sum += sample_reflectance;
        if let Some(values) = &jacobian_values {
            let start = index * jacobian::STATE_COUNT;
            let mut row = jacobian::zero();
            row.copy_from_slice(&values[start..start + jacobian::STATE_COUNT]);
            jacobian::add_scaled(&mut summary.jacobian_sum, row, 1.0);
        }
    }

    let mean_jacobian = jacobian_values.as_ref().map(|_| {
        let mut mean = summary.jacobian_sum;
        for value in &mut mean {
            *value /= sample_count as f64;
        }
        mean
    });
    let product_summary = summary.to_summary(&wavelengths, has_noise_sigma, mean_jacobian);
    Ok(InstrumentGridProduct {
        summary: product_summary,
        wavelengths,
        radiance,
        irradiance,
        reflectance,
        noise_sigma,
        radiance_noise_sigma: Vec::new(),
        irradiance_noise_sigma: Vec::new(),
        reflectance_noise_sigma: Vec::new(),
        jacobian: jacobian_values,
        effective_air_mass_factor: prepared.effective_air_mass_factor,
        effective_single_scatter_albedo: prepared.effective_single_scatter_albedo,
        effective_temperature_k: prepared.effective_temperature_k,
        effective_pressure_hpa: prepared.effective_pressure_hpa,
        gas_optical_depth: prepared.gas_optical_depth,
        cia_optical_depth: prepared.cia_optical_depth,
        aerosol_optical_depth: prepared.aerosol_optical_depth,
        cloud_optical_depth: prepared.cloud_optical_depth,
        total_optical_depth: prepared.total_optical_depth,
        depolarization_factor: prepared.depolarization_factor,
        d_optical_depth_d_temperature: prepared.d_optical_depth_d_temperature,
    })
}
